File: server/src/main/kotlin/pl/edu/agh/eaiib/io/sp/utils/EvaluatorHelper.java
package pl.edu.agh.eaiib.io.sp.utils;

public class EvaluatorHelper {
    public static Integer evaluate(Double accRead, Double min, Double max) {
        if(accRead<=9.81){
            double diff = 9.81 - min;
            if(accRead > 9.81 - diff/10 ){ return 10; }
            else if(accRead > 9.81 - 2*diff/10){ return 9; }
            else if(accRead > 9.81 - 3*diff/10){ return 8; }
            else if(accRead > 9.81 - 4*diff/10){ return 7; }
            else if(accRead > 9.81 - 5*diff/10){ return 6; }
            else if(accRead > 9.81 - 6*diff/10){ return 5; }
            else if(accRead > 9.81 - 7*diff/10){ return 4; }
            else if(accRead > 9.81 - 8*diff/10){ return 3; }
            else if(accRead > 9.81 - 9*diff/10){ return 2; }
            else { return 1; }
        } else {
            double diff = max - 9.81;
            if(accRead < 9.81 + diff/10){ return 10; }
            else if(accRead < 9.81 + 2*diff/10){ return 9; }
            else if(accRead < 9.81 + 3*diff/10){ return 8; }
            else if(accRead < 9.81 + 4*diff/10){ return 7; }
            else if(accRead < 9.81 + 5*diff/10){ return 6; }
            else if(accRead < 9.81 + 6*diff/10){ return 5; }
            else if(accRead < 9.81 + 7*diff/10){ return 4; }
            else if(accRead < 9.81 + 8*diff/10){ return 3; }
            else if(accRead < 9.81 + 9*diff/10){ return 2; }
            else { return 1; }
        }
    }
}

File: server/src/main/kotlin/pl/edu/agh/eaiib/io/sp/utils/EvaluatorHelper.kt
package pl.edu.agh.eaiib.io.sp.utils

fun evaluate(accRead: Double, min: Double, max: Double): Int {
    val conditionEvaluator = getConditionEvaluator(accRead, min, max)
    for (i in 0 until MAX_RATE) {
        val rate = MAX_RATE - i
        if (conditionEvaluator.invoke(i)) {
            return rate
        }
    }

    return MIN_RATE
}

private fun getConditionEvaluator(accRead: Double, min: Double, max: Double): (Int) -> Boolean {
    val diff = Math.abs(if (accRead <= EARTH_ACCELERATION) EARTH_ACCELERATION - min else max - EARTH_ACCELERATION)
    return { i ->
        if (accRead <= EARTH_ACCELERATION) (accRead > EARTH_ACCELERATION - (i + 1) * (diff / MAX_RATE))
        else (accRead < EARTH_ACCELERATION + (i + 1) * (diff / MAX_RATE))
    }
}

private const val EARTH_ACCELERATION = 9.81
private const val MAX_RATE = 10
private const val MIN_RATE = 1